### DISCUSSION TITLE
JsonNode throws "Type must not be Null" on Spring Boot+datastore project when Jackson us upgraded to 2.14.X or higher

### DISCUSSION THREAD

**Discussion by jevanio** [FasterXML/jackson-databind#4470] [Q&A]
Hey folks,
I been working on a spring boot project which use Datastore as backend. On this project, I'm trying to update jackson dependencies from the current version it has (2.13.4) to a newer one (2.16.2) but I been having an issue on a Datastore Entity which use JsonNode as attribute.

I have seen that, after upgrading to jackson 2.14.X or higher, I'm getting an error on bean creation which only says "Type must not be null". After a lot of debugging, and creating a SpringBoot+Datastore project from scratch, I found that if I remove a JsonNode attribute from a Datastore Entity the service starts. I must mention that this JsonNode attribute has been included from long time ago and it was working.

My guess is, after Jackson upgrade to 2.14.X, there's some dependency which needs to be provided to keep it working but I haven't been able to identified which is. I review the Release Notes from 2.14.0 but didn't found anything about it.

Can you point me if there's a breaking change between 2.13.X and higher version or if a similar issue has been identified which can guide me?

**Comment by JooHyukKim** [comment]
You can refer to the [release notes](https://github.com/FasterXML/jackson-databind/blob/2.18/release-notes/VERSION-2.x) to find out what changed

**Reply by JooHyukKim** [reply]
> I'm getting an error on bean creation which only says "Type must not be null".

Also, we need to know how the mentioend bean creation works. Minimal reproduction with only jackson related code and POJOs only would help debugging. Due to nature of this problem `SpringBoot + Datastore + Jackson` we can't tell how it is exactly used..

Folks at Datastore would know...how this called Datastore (library, framework?) is using Jackson, or which Jackson module (there is several)

**Reply by jevanio** [reply]
Thanks for your reply @JooHyukKim 
Let me share with you a demo project which use Spring Boot + Datastore which can help on debugging: https://github.com/jevanio/demoProject
To replicate the error, the only requirement is to replace jackson version defined [on the pom](https://github.com/jevanio/demoProject/blob/master/pom.xml#L18) from 2.13.4 to 2.14.0 (or higher).

About the bean creation, this project just define the datastore entity, datastore repository and autowire it from a service
![image](https://github.com/FasterXML/jackson-databind/assets/1214408/887cfa9a-c41d-4b34-a643-6d2797b07757)

Thank in advance for any advice you can give

**Reply by JooHyukKim** [reply]
As I said above minimal reproduction with only jackson related code and POJOs would allow others to give help, not a whole SpringBoot project which is too much for anyone to spend time to pull and build locally 🤔.

**Reply by jevanio** [reply]
Actually, that's what I did.
If you check the project you will see it only have 4 classes without any logic besides to autowire a datastore repository, as it was shown on the previous comment, and include a JsonNode attribute on a Datastore Entity, as you can see on this image.
![image](https://github.com/FasterXML/jackson-databind/assets/1214408/e5889b4f-fba7-4f47-acd7-3af72d4434ac)

**Reply by cowtowncoder** [reply]
@jevanio Ok this is problematic in that handling of `@Entity` would be outside Jackson. So it'd be necessary to see what data source (?) framework does. I am guessing it perhaps enables "Default Typing" (automatic polymorphic handling), but that is just speculation.

What is needed is to see code for setting up, configuring `ObjectMapper`, and invocation of methods (readValue() / writeValue() ? ).

**Comment by cowtowncoder** [comment]
There are no additional dependencies for `jackson-databind` in 2.14 or later so missing dependency is not the root cause.

So, yeah, what would be needed is a reproduction.

On release notes; changes would be in `jackson-databind` section for:

* https://github.com/FasterXML/jackson/wiki/Jackson-Release-2.14 (2.13 -> 2.14)
* https://github.com/FasterXML/jackson/wiki/Jackson-Release-2.15 (2.14 -> 2.15)
* https://github.com/FasterXML/jackson/wiki/Jackson-Release-2.16 (2.15 -> 2.16)

but it sounds like you may already have gone over some of these.

**Reply by jevanio** [reply]
Actually, I have made a dummy project which with only 4 classes, and without custom logic on the project, it replicate the scenario.
https://github.com/jevanio/demoProject

Please let me know if this help.

Also, based on the Release 2.14 page, I found a reference to JsonNode on https://github.com/FasterXML/jackson-future-ideas/wiki/JSTEP-7 but I'm not sure if this could it be impacting in some way

**Reply by cowtowncoder** [reply]
That link goes to 404 (maybe it's not public?).

JSTEP-7 outlines improvements but none of it should break backwards compatibility.

**Reply by jevanio** [reply]
Thanks for letting me know. I made it public

**Reply by cowtowncoder** [reply]
I can clone the repo; but it looks like:

* Even with default 2.13.4 dependency there's a failure (but different one) -- at least with JDK 17
* Some of transitive dependencies to Jackson modules are newer (2.15.4) -- probably not relevant to issue

Failure itself is from Spring:

```
Caused by: java.lang.reflect.InaccessibleObjectException: Unable to make field final int java.math.BigInteger.signum accessible: module java.base does not "opens java.math" to unnamed module @4ee22985
	at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:354)
	at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:297)
	at java.base/java.lang.reflect.Field.checkCanSetAccessible(Field.java:178)
	at java.base/java.lang.reflect.Field.setAccessible(Field.java:172)
	at org.springframework.util.ReflectionUtils.makeAccessible(ReflectionUtils.java:779)
	at org.springframework.data.mapping.context.AbstractMappingContext$PersistentPropertyCreator.doWith(AbstractMappingContext.java:536)
	at org.springframework.util.ReflectionUtils.doWithFields(ReflectionUtils.java:703)
	at org.springframework.data.mapping.context.AbstractMappingContext.doAddPersistentEntity(AbstractMappingContext.java:422)
	at org.springframework.data.mapping.context.AbstractMappingContext.addPersistentEntity(AbstractMappingContext.java:
```

so maybe I'll need to use older JDK. Will see if things work with 11 or 14.

**Reply by cowtowncoder** [reply]
Ok no, JDK 17 is required.

So, I can see a failure but not the same one as reported.

**Reply by jevanio** [reply]
About that exception, it's something related to Java 17. It require to add as environment variable **JDK_JAVA_OPTIONS=--add-opens java.base/java.math=ALL-UNNAMED**
![image](https://github.com/FasterXML/jackson-databind/assets/1214408/8d6a24ab-79de-452e-b475-a59ead89c9d6)

**Reply by JooHyukKim** [reply]
Just out of curiosity, have u tried inquiring to Datastore team?  @jevanio

**Reply by JooHyukKim** [reply]
I suggest inquiring Stackoverflow or Datastore team, because....

1. There is still no clue on how exactly JsonNode is "used".
2. Jackson does not break on initialization, straight up, so could be something wrong dependency management or misconfiguration

**Reply by cowtowncoder** [reply]
Yeah, at this point there really is little Jackson project can do, not seeing underlying exception, stack trace, or set up of Jackson components.

And even requiring that `--add-opens` seems to be something Datastore team would need to address.

**Reply by jevanio** [reply]
Thanks folks for your comment.
I have already made this same question on Datastore side to get their insights. Really hope they can also guide on this error :)

**Reply by cowtowncoder** [reply]
Good luck @jevanio let's hope they can help!
